// @nut-tree-fork/shared/dist/lib/enums/key.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Key = void 0;
/**
 * The {@link Key} enum represents keys of a standard 105 key US layout keyboard
 */
var Key;
(function (Key) {
    Key[Key["Escape"] = 0] = "Escape";
    Key[Key["F1"] = 1] = "F1";
    Key[Key["F2"] = 2] = "F2";
    Key[Key["F3"] = 3] = "F3";
    Key[Key["F4"] = 4] = "F4";
    Key[Key["F5"] = 5] = "F5";
    Key[Key["F6"] = 6] = "F6";
    Key[Key["F7"] = 7] = "F7";
    Key[Key["F8"] = 8] = "F8";
    Key[Key["F9"] = 9] = "F9";
    Key[Key["F10"] = 10] = "F10";
    Key[Key["F11"] = 11] = "F11";
    Key[Key["F12"] = 12] = "F12";
    Key[Key["F13"] = 13] = "F13";
    Key[Key["F14"] = 14] = "F14";
    Key[Key["F15"] = 15] = "F15";
    Key[Key["F16"] = 16] = "F16";
    Key[Key["F17"] = 17] = "F17";
    Key[Key["F18"] = 18] = "F18";
    Key[Key["F19"] = 19] = "F19";
    Key[Key["F20"] = 20] = "F20";
    Key[Key["F21"] = 21] = "F21";
    Key[Key["F22"] = 22] = "F22";
    Key[Key["F23"] = 23] = "F23";
    Key[Key["F24"] = 24] = "F24";
    Key[Key["Print"] = 25] = "Print";
    Key[Key["ScrollLock"] = 26] = "ScrollLock";
    Key[Key["Pause"] = 27] = "Pause";
    Key[Key["Grave"] = 28] = "Grave";
    Key[Key["Num1"] = 29] = "Num1";
    Key[Key["Num2"] = 30] = "Num2";
    Key[Key["Num3"] = 31] = "Num3";
    Key[Key["Num4"] = 32] = "Num4";
    Key[Key["Num5"] = 33] = "Num5";
    Key[Key["Num6"] = 34] = "Num6";
    Key[Key["Num7"] = 35] = "Num7";
    Key[Key["Num8"] = 36] = "Num8";
    Key[Key["Num9"] = 37] = "Num9";
    Key[Key["Num0"] = 38] = "Num0";
    Key[Key["Minus"] = 39] = "Minus";
    Key[Key["Equal"] = 40] = "Equal";
    Key[Key["Backspace"] = 41] = "Backspace";
    Key[Key["Insert"] = 42] = "Insert";
    Key[Key["Home"] = 43] = "Home";
    Key[Key["PageUp"] = 44] = "PageUp";
    Key[Key["NumLock"] = 45] = "NumLock";
    Key[Key["NumPadEqual"] = 46] = "NumPadEqual";
    Key[Key["Divide"] = 47] = "Divide";
    Key[Key["Multiply"] = 48] = "Multiply";
    Key[Key["Subtract"] = 49] = "Subtract";
    Key[Key["Tab"] = 50] = "Tab";
    Key[Key["Q"] = 51] = "Q";
    Key[Key["W"] = 52] = "W";
    Key[Key["E"] = 53] = "E";
    Key[Key["R"] = 54] = "R";
    Key[Key["T"] = 55] = "T";
    Key[Key["Y"] = 56] = "Y";
    Key[Key["U"] = 57] = "U";
    Key[Key["I"] = 58] = "I";
    Key[Key["O"] = 59] = "O";
    Key[Key["P"] = 60] = "P";
    Key[Key["LeftBracket"] = 61] = "LeftBracket";
    Key[Key["RightBracket"] = 62] = "RightBracket";
    Key[Key["Backslash"] = 63] = "Backslash";
    Key[Key["Delete"] = 64] = "Delete";
    Key[Key["End"] = 65] = "End";
    Key[Key["PageDown"] = 66] = "PageDown";
    Key[Key["NumPad7"] = 67] = "NumPad7";
    Key[Key["NumPad8"] = 68] = "NumPad8";
    Key[Key["NumPad9"] = 69] = "NumPad9";
    Key[Key["Add"] = 70] = "Add";
    Key[Key["CapsLock"] = 71] = "CapsLock";
    Key[Key["A"] = 72] = "A";
    Key[Key["S"] = 73] = "S";
    Key[Key["D"] = 74] = "D";
    Key[Key["F"] = 75] = "F";
    Key[Key["G"] = 76] = "G";
    Key[Key["H"] = 77] = "H";
    Key[Key["J"] = 78] = "J";
    Key[Key["K"] = 79] = "K";
    Key[Key["L"] = 80] = "L";
    Key[Key["Semicolon"] = 81] = "Semicolon";
    Key[Key["Quote"] = 82] = "Quote";
    Key[Key["Return"] = 83] = "Return";
    Key[Key["NumPad4"] = 84] = "NumPad4";
    Key[Key["NumPad5"] = 85] = "NumPad5";
    Key[Key["NumPad6"] = 86] = "NumPad6";
    Key[Key["LeftShift"] = 87] = "LeftShift";
    Key[Key["Z"] = 88] = "Z";
    Key[Key["X"] = 89] = "X";
    Key[Key["C"] = 90] = "C";
    Key[Key["V"] = 91] = "V";
    Key[Key["B"] = 92] = "B";
    Key[Key["N"] = 93] = "N";
    Key[Key["M"] = 94] = "M";
    Key[Key["Comma"] = 95] = "Comma";
    Key[Key["Period"] = 96] = "Period";
    Key[Key["Slash"] = 97] = "Slash";
    Key[Key["RightShift"] = 98] = "RightShift";
    Key[Key["Up"] = 99] = "Up";
    Key[Key["NumPad1"] = 100] = "NumPad1";
    Key[Key["NumPad2"] = 101] = "NumPad2";
    Key[Key["NumPad3"] = 102] = "NumPad3";
    Key[Key["Enter"] = 103] = "Enter";
    Key[Key["LeftControl"] = 104] = "LeftControl";
    Key[Key["LeftSuper"] = 105] = "LeftSuper";
    Key[Key["LeftWin"] = 106] = "LeftWin";
    Key[Key["LeftCmd"] = 107] = "LeftCmd";
    Key[Key["LeftAlt"] = 108] = "LeftAlt";
    Key[Key["LeftMeta"] = 109] = "LeftMeta";
    Key[Key["RightControl"] = 110] = "RightControl";
    Key[Key["RightSuper"] = 111] = "RightSuper";
    Key[Key["RightWin"] = 112] = "RightWin";
    Key[Key["RightAlt"] = 113] = "RightAlt";
    Key[Key["RightCmd"] = 114] = "RightCmd";
    Key[Key["RightMeta"] = 115] = "RightMeta";
    Key[Key["Space"] = 116] = "Space";
    Key[Key["Menu"] = 117] = "Menu";
    Key[Key["Fn"] = 118] = "Fn";
    Key[Key["Left"] = 119] = "Left";
    Key[Key["Down"] = 120] = "Down";
    Key[Key["Right"] = 121] = "Right";
    Key[Key["NumPad0"] = 122] = "NumPad0";
    Key[Key["Decimal"] = 123] = "Decimal";
    Key[Key["Clear"] = 124] = "Clear";
    Key[Key["AudioMute"] = 125] = "AudioMute";
    Key[Key["AudioVolDown"] = 126] = "AudioVolDown";
    Key[Key["AudioVolUp"] = 127] = "AudioVolUp";
    Key[Key["AudioPlay"] = 128] = "AudioPlay";
    Key[Key["AudioStop"] = 129] = "AudioStop";
    Key[Key["AudioPause"] = 130] = "AudioPause";
    Key[Key["AudioPrev"] = 131] = "AudioPrev";
    Key[Key["AudioNext"] = 132] = "AudioNext";
    Key[Key["AudioRewind"] = 133] = "AudioRewind";
    Key[Key["AudioForward"] = 134] = "AudioForward";
    Key[Key["AudioRepeat"] = 135] = "AudioRepeat";
    Key[Key["AudioRandom"] = 136] = "AudioRandom";
})(Key || (exports.Key = Key = {}));
